# 03. Generative Language Models

## Paper requirements

3 step of the reproduction of a DNN Scientific Paper(SP):
1. Choose a SP that needs to be published after 1st of January 2023
2. Do a resume of the main parts of the paper
3. Reproduce in details part of the paper code

You can do it all at once, but strongly recommended to do it on stages.

## GELU (Gaussian Error Linear Units)

Problem wtih RELU. If the activation functions is negative it outputs 0, if it outputs 0 then the gradient is 0, then the unit wont be doing any learning on that iteration.

If you notice that the DNN is learning really slow and you are using RELU, that could be the problem, you might have many units that are not learning due to negative activation going to 0.\
Quick fix is to use leaky_relu that puts a 0,0001 instead of 0.

<img src="Images/gelu.png" width=500>

(Gemini)\
**GELU** is a high-performance, non-linear activation function commonly used in Transformer architectures like **BERT** and **GPT-2**.

#### 1. The Mathematical Definition
The function is defined as:
$$GELU(x) = x\Phi(x)$$

Where:
* $x$ is the input to the activation.
* $\Phi(x)$ is the **Standard Gaussian Cumulative Distribution Function**.

#### 2. Probabilistic Intuition
GELU can be thought of as a smoother, probabilistic version of ReLU. Instead of a hard "gate" that shuts off at zero, it weights the input by its probability under a normal distribution:
$$x \cdot P(X \le x)$$

#### 3. Why use GELU over ReLU?
* **Smoothness:** Unlike ReLU, which has a sharp "elbow" at zero (making it non-differentiable at that exact point), GELU is smooth everywhere.
* **Vanishing Gradients:** Because GELU allows for a small, non-zero gradient even for slightly negative values, it helps mitigate the "dying ReLU" problem during backpropagation.
* **Transformer Performance:** Research has shown that the curvature of GELU allows models to converge faster and reach higher accuracy in complex NLP tasks.

> **Note:** Because the Gaussian CDF is computationally expensive to calculate exactly, most deep learning frameworks use a fast approximation:
> $$0.5x \left(1 + \tanh\left[\sqrt{2/\pi}(x + 0.044715x^3)\right]\right)$$

## Root Mean Squared Propagation (RMSprop) Optimizer

RMSProp is an adaptive learning rate optimization algorithm for training deep neural networks. It tackles the vanishing/exploding gradient problem by normalizing gradients using a moving average of their squared magnitudes, allowing for faster convergence in non-stationary objectives and reducing vertical oscillations. 

In [2]:
from transformers import GPT2LMHeadModel, AutoConfig, AutoTokenizer

/home/alpine/miniconda3/envs/dl/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
config = AutoConfig.from_pretrained("gpt2")

In [3]:
gpt2 = GPT2LMHeadModel(config)

In [4]:
gpt2

GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D(nf=2304, nx=768)
          (c_proj): Conv1D(nf=768, nx=768)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D(nf=3072, nx=768)
          (c_proj): Conv1D(nf=768, nx=3072)
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)

## Dropout

```
(attn_dropout): Dropout(p=0.1, inplace=False)
(resid_dropout): Dropout(p=0.1, inplace=False)
```
We can see these drop layers on top in the GPT architecture. What are they and what do they do?

Dropout is basically regularization. It is the DNN's way of doing that.

```p=0.1``` means to drop 10% of the neurons in the layers it applies to during that training step. Each training step they are different.

> Dropout is applied **only during training**, not during inference!

<img src="Images/dropout.png" width=500>

In [1]:
from keras.layers import Dense, Dropout
from keras.regularizers import L1L2, L1, L2

2026-02-05 19:14:29.109977: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [6]:
Dense(units = (20, 200), kernel_regularizer = L1L2(), bias_regularizer = L1() , activity_regularizer = L2() )

I0000 00:00:1770207950.816064   68269 gpu_device.cc:2020] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 6100 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 2060 SUPER, pci bus id: 0000:08:00.0, compute capability: 7.5


<Dense name=dense, built=False>

This is the regularization that we have studied in the ML course. L2 for example:

$$ J + \lambda * ||w||^2_2 $$

we add the weights of the model to the loss function with some norm

The cool thing is we can add regularization to each part of the model:
- kernal: the weights
- bias - bias
- activity - the result

This yet powerful is also dangerous. We have to very carefully tune it. The model can jump between high bias and high variance with a tiny lambda change.

It is much easier to just use a Dropout(). It is not a layer itself, it is just applied on the layer before it.

In [9]:
Dense(units = (20, 200), kernel_regularizer = L1L2(), bias_regularizer = L1() , activity_regularizer = L2() )
Dropout(rate=0.1)

<Dropout name=dropout, built=True>

## Norm

Normalization, what a shocker!

We want out inputs to be small numbers. It speeds up learning and computation

(Gemini)
###  Why Normalization Matters

Normalization (like **LayerNorm** in GPT-2) ensures that the values flowing through the model stay within a manageable range.

* **Learning Speed:** Smoothes the loss landscape, allowing the optimizer to reach the "bottom" (minimal error) in fewer iterations.
* **Stability:** Mitigates the **Exploding/Vanishing Gradient** problem by re-centering data at every layer.
* **Higher Learning Rates:** Because the math is more stable, we can train faster without the model "breaking."

**The Math:**
For an input $x$, we calculate the mean ($\mu$) and variance ($\sigma^2$) and transform $x$ as:
$$\hat{x} = \frac{x - \mu}{\sqrt{\sigma^2 + \epsilon}}$$
*(where $\epsilon$ is a tiny constant to prevent division by zero)*

> **Note:** GPT-2 uses **LayerNorm**, which normalizes across the hidden features of a single token, rather than **BatchNorm**, which looks at multiple sentences. This is crucial for consistent performance during inference.

### Normalization Comparison: Layer vs. Batch

| Feature | Layer Normalization (LN) | Batch Normalization (BN) |
| :--- | :--- | :--- |
| **Direction** | Across all features of **one** sample. | Across all samples of **one** feature. |
| **Batch Size** | Independent (works with Batch Size = 1). | Highly dependent (needs large batches). |
| **Use Case** | **Transformers**, RNNs, NLP. | CNNs, Computer Vision. |
| **Inference** | Same behavior as training. | Uses "running statistics" from training. |

#### 1. LayerNorm: The Transformer Choice
LayerNorm treats each sentence as its own mini-universe. It calculates the mean and variance of all 768 hidden units for a single token. This is crucial for Transformers because it ensures that the "self-attention" scores aren't affected by other unrelated sentences in the same training batch.

#### 2. BatchNorm: The CNN Classic
BatchNorm looks at "Feature #5" across 32 different images at once. While powerful for images, it fails in NLP because sentences have different lengths and meanings; forcing them to share statistics across a batch often "muddies" the specific context of a single sentence.

We use LayerNorm, because it is faster. We don't have to wait for the whole batch to be processed to apply LayerNorm, whereas with BatchNorm we do. It is just faster.

There are other normalization: LazyBatch, GroupNorm, RMSNorm etc.

In general, the model does:

1. Tokenization
2. Feature extraction (the attension block)
3. Some head (decoder) that performs a specific task

Tokenization could be a whole separate problem in itself. There are many things we can do/tweak to that process:

- Pre tokenization: Don't allow a token to contain the end of a work, then space, then the beginning of another word. First we split on space, then we tokenize.
- Remove emoticons
- Remove numbers
- Different batch sizes
- etc.

Tokenization is a different process from Attension, but they are trained together

Example theoretical questions:
- Image captioning - what is a good loss function for this problem? why choose it? What metrics can we use? Would you use attension with 2 or 15 heads?

## Language Model Architecture

### Encoder / Decoder Architecture

- Encode - (understand) encode the input into a **latent representation**
- Decode - use the representation to decode (geenrate), possibly to adifferent type of output, e. g., an image

<img src="Images/encoder_decoder.png" width=500>

### Autoencoders

At first glance it looks like they aren't doing anything, but they do alot actually.

- The input and output dimensions need to be the same
- The loss function needs to be $ J = MSE(\hat{y}, x) $
- Have an encoder/decoder architecture in the hidden layers, like a compression decompression

<img src="Images/autoencoder.png" width=500>

**We don't even need to have labels, we are just reconstructing our inputs**

But we still have a regression loss function, so this is called **Self-supervised learning**

Best way to deal with huge amounts of data

Advantages:
- Dimensionality reduction
- Denoising

### Dimensionality reduction

We use the decoder just during the training, we can then throw it away and we are left with an encoder that we are sure compresses the data into a smaller latent (hidden) representation!!!

Good for tabulated data. The process is PCA essentially, by removing the decoder.
### Denoiser
Not exactly autoencoderm, since it outputs a different thing, but it is the same concept

<img src="Images/denoiser.png" width=500>

In [8]:
from transformers import AutoTokenizer


tokenizer = AutoTokenizer.from_pretrained("gpt2")

In [9]:
tokenizer.eos_token

'<|endoftext|>'

Tokens have this syntax <| ... |>

When doing tokenization, it is very important not to split a token that the tokenizer uses (like <|endoftext|>) in to tokens

One of the way to attack a model is to give it a token it is using (systemic token)

In [10]:
tokenizer.encode("A black cat in a black box.", return_tensors="pt")

tensor([[  32, 2042, 3797,  287,  257, 2042, 3091,   13]])

In [11]:
tokenizer.decode([  32, 2042, 3797,  287,  257, 2042, 3091,   13])

'A black cat in a black box.'

### Lets [compare](https://modal.com/blog/gpt-oss-arch) the GPT-2 architecture to GPT OSS (open source)

GTP OSS:
- Doesn't use Dropout as it is not needed turns out. Attension and Droupout don't do well together
    - When training a foundational model like GPT on huge amounts of data, we want it to overfit on the data, Dropout is against that.
    - Now if we fine-tune it after, we don't want that and we can use it there.
- Uses SWISH over GELU activation, looks almost the same, but faster to compute
- Sliding window attension. The length of the window doesn't need to be fixed
    - <img src="Images/sliding_window.png" width=500> 
    - Convolutions are ver short and fast, but attensions is very big and slow. We can use something inbetween.
    - We lose a tiny amount of accuracy for less computation. That was the idea of this optimization
    - Negatives, are that it loses context on long contexts
- RMSNorm over LayerNorm
   - Less accurate than LayerNorm, but much faster to compute
- RoPE(Rotary positional Embeddings) over Positional Embeddings
    - It is a new way of giving it many tokens.
    - <img src="Images/rope.png" width=500> 
    - Each token gets 2 positional embeddings, not just 1 (we see we put them in a circle (rotary) and get their sin and cos positions in it, giving us 2 dimensions.
    - RoPE rotates a token’s query and key vectors where the angle is determined by the token’s position. The rotation happens on a complex number plane; accordingly, RoPE encodes relative position into the attention computation without having to separately track embeddings.
    - Allows to handle longer contexts
- Mixture-of-Experts (MoE)
    - We have an ensamble of models, but we don't call all of them
    - <img src="Images/moe.png" width=500> 
    - Router chooses K experts/layers to call during inference and when K < N (meaning we don't use all of them) it will work faster (it performs classification on who is the most likely to do a good job)
    - We traing them all, but use a subset on inference
    - The router decides which layers are better for the specific task on inference
    - We can also see a skip connections (residual network) where we skip the layers and give the input directly to the output

In [12]:
model = GPT2LMHeadModel.from_pretrained("gpt2")

Loading weights: 100%|███████████████████| 148/148 [00:00<00:00, 541.91it/s, Materializing param=transformer.wte.weight]
GPT2LMHeadModel LOAD REPORT from: gpt2
Key                  | Status     |  | 
---------------------+------------+--+-
h.{0...11}.attn.bias | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [13]:
model.eval()

GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D(nf=2304, nx=768)
          (c_proj): Conv1D(nf=768, nx=768)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D(nf=3072, nx=768)
          (c_proj): Conv1D(nf=768, nx=3072)
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)

Often you can use what you have learned with the model and use some strategy for generation

### Generative Models Intuition

1. Supervised learning: $ (x, y) \Rightarrow \hat{y} = f(x): y \approx \hat{y}$
2. Unsupervised learning: $ x \Rightarrow f(x): f $ describes structure

Generative models use something like unsupervised learning

- Learn the distribution of training data $ p_{train} (x) = f(x) $
- Make the model output new data $ p_{model} (x) $ "similar" to  $ p_{train} $
- In other words: after training, make a model: $ \hat{y} = f(\vec{0}) $

<img src="Images/gen_models.png" width=500> 

We learn the original distribution and then we sample from it.

In [33]:
inputs = tokenizer("To write a short horror story you must ", return_tensors="pt")

In [26]:
model = GPT2LMHeadModel.from_pretrained("gpt2").eval()

Loading weights: 100%|███████████████████| 148/148 [00:00<00:00, 544.67it/s, Materializing param=transformer.wte.weight]
GPT2LMHeadModel LOAD REPORT from: gpt2
Key                  | Status     |  | 
---------------------+------------+--+-
h.{0...11}.attn.bias | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [34]:
result = model.generate(**inputs, max_new_tokens = 100, do_sample = False, num_beams = 1) # do_sample = False, num_beams = 1 is for greedy

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


In [35]:
result

tensor([[2514, 3551,  257, 1790, 9961, 1621,  345, 1276,  220, 1849, 1350, 1498,
          284, 3551,  257, 1621,  326,  318,  407,  655,  257, 9961, 1621,  475,
          257, 1621,  326,  318,  407,  655,  257, 9961, 1621,  475,  257, 1621,
          326,  318,  407,  655,  257, 9961, 1621,  475,  257, 1621,  326,  318,
          407,  655,  257, 9961, 1621,  475,  257, 1621,  326,  318,  407,  655,
          257, 9961, 1621,  475,  257, 1621,  326,  318,  407,  655,  257, 9961,
         1621,  475,  257, 1621,  326,  318,  407,  655,  257, 9961, 1621,  475,
          257, 1621,  326,  318,  407,  655,  257, 9961, 1621,  475,  257, 1621,
          326,  318,  407,  655,  257, 9961, 1621,  475,  257, 1621,  326,  318,
          407]])

In [36]:
tokenizer.decode(result[0])

'To write a short horror story you must \xa0be able to write a story that is not just a horror story but a story that is not just a horror story but a story that is not just a horror story but a story that is not just a horror story but a story that is not just a horror story but a story that is not just a horror story but a story that is not just a horror story but a story that is not just a horror story but a story that is not just a horror story but a story that is not'

The model isnt trained to follow instructions.

### Multi-nomial sampling

Model doesn't just take the most probable token but also others that are not as probable.

#### Top-k sampling (number)
Top k number of samples

#### Top-p sampling (probability)

Top probability samples
`top_p=0.8` Give me the top however many samples that commulatively add up to 80% probability

We can see the options to use multinomial sampling here:

<img src="Images/sampling.png" width=800> 

In [37]:
inputs = tokenizer("To write a short horror story you must ", return_tensors="pt")

tokenizer.decode(model.generate(**inputs, max_new_tokens = 100, do_sample = True, num_beams = 1, top_k = 10)[0])

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


"To write a short horror story you must \xa0do a bit of work with this. First of all, I want to make sure that the story is not too complex. You may want to add in the following elements to this story, like the fact that some people are going to love it (I have a strong suspicion that they will, but I'm not sure I want to spoil the fun). I also want to make sure that the story is not overly dramatic (like I do with my stories). I want the reader to be able"

We are saying pick the top 10 most probable next tokens and choose 1 of them

We can see the result is better.

This allows us to control how much it hallucinates

In [40]:
inputs = tokenizer("To write a short horror story you must ", return_tensors="pt")

tokenizer.decode(model.generate(**inputs, max_new_tokens = 100, do_sample = True, num_beams = 1, top_k = 10, temperature = 0.9)[0])

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


'To write a short horror story you must \xa0be a\xa0good writer and a\xa0good storyteller,\xa0and that includes a good storyteller with good writing skills (like a good writer who writes well and is able to do so well).\xa0 The\xa0best\xa0storyteller\xa0is a\xa0good\xa0writer\xa0that is a good storyteller.\xa0\xa0It is also important to know a good storyteller.\xa0 A good storyteller is someone who is able to write an excellent story, and who'

#### Temperature

Is the delimiter in the softmax function.

<img src="Images/temperature.png" width=400> 

$ t $ is between $ (0, 1) $

<img src="Images/temperature_diff.png" width=600>

The more the temperature rises the more the probabilites are flattened out and the same

We can use `StoppingCriteria` from huggingFace to stop the generation earlier 

These are called Decoding Strategies:
- Greedy sampling
- Temperature
- Top-k sampling
- Top-p sampling
- Beam search

#### Beam search

Imagine a flashlight, beam is the width of the flashlight

We look ahead of every combination of words. Beam is the param that determines the leaf number. In the picture it is num_beams=3 with depth 3

Hyperparam: `num_beams` controls this process

<img src="Images/beam.png" width=600> 

Problem is it is a lot slower.

In [43]:
inputs = tokenizer("To write a short horror story you must ", return_tensors="pt")

tokenizer.decode(model.generate(**inputs, max_new_tokens = 100, do_sample = True, num_beams = 5, temperature = 0.7, top_k = 10)[0])

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


'To write a short horror story you must \xa0be able to tell the story in a way that makes sense to the reader. You must \xa0be able to tell the story in a way that makes sense to the reader. You must \xa0be able to tell the story in a way that makes sense to the reader. You must \xa0be able to tell the story in a way that makes sense to the reader. You must \xa0be able to tell the story in a way that makes sense to the reader. You must '

## LLM Training Steps

1. Pre-train
    - Self-supervised - no labels, the idea is to learn the data itself, like feature extraction, to learn which tokens go with which other tokens
    - Ingest a lot of data (trillions of tokens)
    - No particular task
3. Fine-tune
    - Usually supervised learning - with labels
    - Specialize the model to do something specific
    - (Get the base model from the pre-train and specify it into something)

ChatGPT, Gemini etc. are fine-tuned to be instruction models, or chats. Based on the input (question or instruction) they generate and output.

This is called instruction tuning

### DPO (Direct preference optimization)

Preference tuning fine_tunes LLMs to align with human choices by training directly on paired preference data (a good response vs a bad one)

It uses a specific loss function will force the bad answer to gie a higher loss result than the good result's loss 

<img src="Images/dpo.png" width=600> 

## Other Generative Models (VAE, GAN, diffusion)

### VAE - Variational Autoencoders

In an autoencoder architecture, we know we can remove the decoder and we are left with a useful encoder that does dim reduction.\
But can we remove the decoder?

Technically yes, but what do we give it? The things in the vector between the enc and decoder are a mystery. Here comes VAE.

<img src="Images/vae.png" width=600> 

During training we add a rule that the vector between the encoder and the decorder should be Z scored. Mean = 0, STD = 1

Now we know what the distribution of that vector is and we can sample other vectors from it at random. Now we don't need the encoder

#### VAE loss

<img src="Images/vae_loss.png" width=600> 

KLD(divergence) - entropy metric, how much one distribution is similar to another

We can see we added the rule for the output to look like Gausian distribution

Problem is we can't really tweak the generation in our favour like in LLMs, with text based instructions.

These are usually used for Density Modelling - to understand the original data distribution and generating auto generated data

### GAN (Generative Adversarial Networks)

Models should expect malicious inputs like: how to make a bomb etc.

(Gemini)

A **Generative Adversarial Network (GAN)** is a system where two neural networks compete against each other in a high-stakes game of "Counterfeiter vs. Detective."

---

### The Two Players

1. **The Generator (The Counterfeiter)**
   * **Goal:** Create fake data (like images) that is indistinguishable from real data.
   * **Process:** It starts with random noise and transforms it into a structured output.

2. **The Discriminator (The Detective)**
   * **Goal:** Correctly identify whether an input is **Real** (from the dataset) or **Fake** (from the Generator).
   * **Process:** It acts as a binary classifier, outputting a probability between 0 and 1.

---

### How It Works
The two networks are trained simultaneously in a "minimax" game:
* The **Discriminator** learns to spot the flaws in the fake data.
* The **Generator** uses the Discriminator's feedback to improve its "forgeries."



### The Objective Function
Mathematically, the GAN tries to reach a state where the Generator has perfectly modeled the real data distribution. This is often represented by the value function $V(D, G)$:

$$\min_{G} \max_{D} V(D, G) = \mathbb{E}_{x \sim p_{data}(x)}[\log D(x)] + \mathbb{E}_{z \sim p_{z}(z)}[\log(1 - D(G(z)))]$$

* **Equilibrium:** The training is successful when the Discriminator can no longer tell the difference, resulting in a probability of **0.5** (50%) for all inputs.

Once the training is done we can use the Generator part of it to generate images.

They used to be very popular, but now not so much. Still really good for images. Mostly used for interpretationo r finding failure modes (situations our models can fail in) 

### Diffusion-based Models

Comes from diffusion in physics:
- A process that happens when we mix 2 things
- <img src="Images/diffusion.gif" width=300> 
- We mix 2 liquids, wait a bit and finally get a homegenous liquid

(Gemini)

Diffusion models are a class of generative models that create data by learning to reverse a process of gradual degradation. They are based on two symmetrical phases: the Forward Process and the Reverse Process.

<img src="Images/diff_model.gif" width=300> 

#### The Forward Process (Diffusion)
The forward process, also known as the diffusion process, systematically converts structured data into random noise. 

* **Mechanism:** Given a data point $x_0$, a small amount of Gaussian noise is added over $T$ successive steps. This forms a Markov chain where each step $x_t$ depends only on the previous step $x_{t-1}$.
* **Mathematical Definition:** The transition is defined as:
  $$q(x_t | x_{t-1}) = \mathcal{N}(x_t; \sqrt{1 - \beta_t} x_{t-1}, \beta_t \mathbf{I})$$
  where $\beta_t$ is a variance schedule.
* **Result:** As $t \to T$, the data distribution $x_T$ becomes indistinguishable from pure isotropic Gaussian noise. This process is fixed and requires no training.



---

#### The Reverse Process (Denoising)
The generative power of the model lies in its ability to undo the forward process. Since the exact reverse distribution $q(x_{t-1} | x_t)$ is unknown, a neural network is trained to approximate it.

* **The Objective:** The model, typically a **U-Net** or **Transformer**, is trained to predict the noise $\epsilon$ that was added at any given timestep $t$. 
* **Training Function:** The loss function simplifies to the Mean Squared Error (MSE) between the **actual noise added** and the **noise predicted** by the network:
  $$L = \mathbb{E}_{x_0, \epsilon, t} [ \| \epsilon - \epsilon_\theta(x_t, t) \|^2 ]$$
* **Inference (Sampling):** To generate a new image, the model starts with random noise $x_T$ and iteratively applies the learned denoising function to step backward from $t=T$ to $t=0$, revealing the final data point.

---

<img src="Images/diff_architecture.png" width=700> 

<img src="Images/diff_usage.png" width=700> 

#### Conditioning
Conditioning is the mechanism that allows you to "steer" Stable Diffusion. Without it, the model would simply generate a random, high-quality image from its training distribution.

In technical terms, conditioning injects external information (like text, images, or depth maps) into the denoising process to guide the model toward a specific output.

- Pass text prompt at the beginning
- Give the loss function $\hat{n}$(t| no prompt) - $\hat{n}$(t|prompt). This way we say we prefer the prompt + image example much more over the no prompt + image 
#### Key Components

* **Noise Schedule:** A predefined sequence of $\beta$ values that determines how quickly the data is "destroyed."
* **Time Embeddings:** Because the model uses the same weights for every step, it needs a "timestamp" (time embedding) to know whether it is looking at pure noise (near $T$) or refining fine details (near $0$).
* **U-Net Architecture:** Most diffusion models use a U-Net with ResNet blocks and Self-Attention layers to capture both local textures and global structure during the denoising steps.